### PR TITLE
googleapis: use modern CMake integrations

### PR DESCRIPTION
**Pull request by jcar87**
Specify library name and version:  **googleapis/all**

### Summary of changes

This PR updates the `googleapis` recipe to use functionality that is compatible with both Conan 1.x and Conan 2.0

* Use `CMakeDeps` and `CMakeToolchain` instead of legacy integrations
* Add transitive headers trait to the protobuf dependency - this is for Conan 2.0. `.pb.h` headers inside this package have transitive public header dependencies on Protobuf
* Minor changes to use new functions from `conan.tools`
* Add `resdirs` cpp_info attribute - it needs to be explicitly set in Conan 2.0 as there is no default.
* Add `package_type` attribute

### Shorten build paths in CMake-generated files:
* Modify the logic in the CMake code generation such that the target names _do not_ start with `google_` or end with `_proto` - this is in an attempt to get cmake to generate shorter vcxproj files, such that the path to the build-time-only `.dir` directory and all temporary files generated by MSBuild have shorter names, to work around the following error:
```
C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\VC\VCTargets\Microsoft.CppBuild.targets(321,5): error MSB3491: Could not write lines to file "google_cloud_beyondcorp_clientconnectorservices_v1_clientconnectorservices_proto.dir\Release\google_c.96C68757.tlog\google_cloud_beyondcorp_clientconnectorservices_v1_clientconnectorservices_proto.lastbuildstate". The specified path, file name, or both are too long. The fully qualified file name must be less than 260 characters, and the directory name must be less than 248 characters. [C:\J\w\prod\BuildSingleReference@5\s\495e34\1\build\google_cloud_beyondcorp_clientconnectorservices_v1_clientconnectorservices_proto.vcxproj]
```

The _file_ names of the built libraries are preserved, as well as the target names that we are parsing in `package_info`.

### CODE REVIEW

**Review by franramirez688 — DISMISSED** [review]

File: recipes/googleapis/all/conanfile.py [hunk 1]
```diff
@@ -38,26 +38,28 @@ class GoogleAPIS(ConanFile):
...
-        self.copy("CMakeLists.txt")
-        for patch in self.conan_data.get("patches", {}).get(self.version, []):
-            self.copy(patch["patch_file"])
+        copy(self, "CMakeLists.txt", src=self.recipe_folder, dst=os.path.join(self.export_sources_folder, "src"))
+        export_conandata_patches(self)
 
     def source(self):
         get(self, **self.conan_data["sources"][str(self.version)], destination=self.source_folder, strip_root=True)
-        self._patch_sources()
+        apply_conandata_patches(self)
```
Comment (prince-chrismc) [inline recipes/googleapis/all/conanfile.py:46]: Patches should be applied in the build method 🤔

**Comment by jcar87** [comment]
```
C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\VC\VCTargets\Microsoft.CppBuild.targets(321,5): error MSB3491: Could not write lines to file "google_cloud_beyondcorp_clientconnectorservices_v1_clientconnectorservices_proto.dir\Release\google_c.96C68757.tlog\google_cloud_beyondcorp_clientconnectorservices_v1_clientconnectorservices_proto.lastbuildstate". The specified path, file name, or both are too long. The fully qualified file name must be less than 260 characters, and the directory name must be less than 248 characters. [C:\J\w\prod\BuildSingleReference@5\s\495e34\1\build\google_cloud_beyondcorp_clientconnectorservices_v1_clientconnectorservices_proto.vcxproj]
  Running cpp protocol buffer compiler on C:/J/w/prod/BuildSingleReference@5/s/495e34/1/src/google/cloud/beyondcorp/clientgateways/v1/client_gateways_service.proto

  ```

**Review by RubenRBS — APPROVED** [review]

**Review by RubenRBS — APPROVED** [review]
:)

**Review by prince-chrismc — APPROVED** [review]

**Review by uilianries — APPROVED** [review]
LGTM

**Comment by kbarry-aurora** [comment]
Is it possible that this broke `grpc-proto`?

```text
ERROR: grpc-proto/cci.20220627: Error in build() method, line 97
	cmake = self._configure_cmake()
while calling '_configure_cmake', line 63
	cmake.definitions["GOOGLEAPIS_PROTO_DIRS"] = self.dependencies["googleapis"].cpp_info.resdirs[0].replace("\\", "/")
	IndexError: list index out of range
The command '/bin/sh -c conan install --build=missing --settings="compiler.cppstd=20" ..' returned a non-zero code: 1
```

**Comment by prince-chrismc** [comment]
`resdir` is correctly being set but it's now on one component... maybe!

https://github.com/conan-io/conan-center-index/blob/master/docs/consuming_recipes.md#isolate-your-project-from-upstream-changes

**Pull request merged**